keps/sig-windows/2802-identify-windows-pods-apiserver-admission/README.md

```diff
@@ -0,0 +1,401 @@
+  // enable the IdentifyPodOS feature. This is used to help identifying the
+  // OS authoritatively during the API server admission.
+  // +optional
+  OS *OS
+}
+
+// OS has information on the type of OS. 
+// We're making this a struct for possible future expansion.
+type OS struct {
+  // Name of the OS. The values supported are available at:
+  // https://github.com/opencontainers/runtime-spec/blob/master/config.md#platform-specific-configuration
+  Name string
+}
+
+```
+
+With the above change, all the admission plugins which validate or mutate the pod can 
+identify the pod's OS authoritatively and can act accordingly. As of now, PodSecurityAdmission
+plugin is the only admission plugin can leverage this field on the pod spec. In addition, API validators like ValidatePod and ValidatePodUpdate should be modified. In future, we can
+have a validating admission plugin for Windows pods as Linux and Windows host capabilities are
+different and not all Kubernetes features are supported on Windows host.
+
+
+### User Stories (Optional)
+
+
+#### Story 1
+
+As a Kubernetes cluster administrator, I want appropriate security contexts to be 
+applied to my Windows Pods along with Linux pods
+
+#### Story 2
+
+As a Kubernetes cluster administrator, I want to use my own admission webhook for pods
+based on the OS it intends to run on.
+
+### Notes/Constraints/Caveats (Optional)
+
+Since the OS field is optional in the pod spec, we can expect the users to omit this field
+when submitting pod to API server. In this scenario, we expect admission plugins to treat the pod
+the way it is being treated now.
+
+### Risks and Mitigations
+
+The primary risk is a bug in the implementation of the admission plugins that validate or mutate
+based on the OS field in the pod spec. The best mitigation for that scenario is unit testing 
+when this featuregate is enabled and disabled.
+Additionally, there may be some end-user confusion on the functional consequences of setting the new OS field, given that it is optional.
+
+## Design Details
+
+### Changes to Kubelet
+
+Apart from the above API change, we intend to make the following changes to Kubelet:
+- Kubelet should reject admitting pod if the kubelet cannot honor the pod.Spec.OS.Name. For instance, if the OS.Name does not match the host os.
+- Kubelet should reject admitting pods if the pod contains a nodeSelector with label `kubenetes.io/os` with a value 
+not matching `runtime.GOOS`. We can use this check as `kubernetes.io/os` value is supposed to be filled from `runtime.GOOS`. 
+
+Having the above checks in kubelet helps in the following conditions:
+- If scheduler has been bypassed by setting the nodeName directly on the pod spec.
+- A malicious user or a user unknowingly with RBAC to just bind pod to node, can associate pod to node there by bypassing the scheduler.
+
+As of now, Kubelet doesn't reconcile the value of `kubernetes.io/os`. So, it'd be nice to have the reconciliation as part of this change.
```

Comment:
this seems like a fine change to make, but I don't think it's related to this KEP any more (definitively identifying the OS of a pod)

Reply:
I agree with @liggitt that its good for kubelet to reconcile the label, but it is distinct.

The kubelet changes described in this section lgtm.

Reply:
Removed it altogether